packages/wallet/core/src/signers/session-manager.ts

```diff
@@ -283,6 +285,48 @@ export class SessionManager implements SapientSigner {
     }
   }
 
+  async signEnvelope(
+    wallet: Address.Address,
+    chainId: number,
+    payload: Payload.Parented,
+    imageHash: Hex.Hex,
+    guard?: GuardSigner,
+  ): Promise<(Envelope.SapientSignature | Envelope.Signature)[]> {
+    const signature = await this.signSapient(wallet, chainId, payload, imageHash)
+    const sapientSignature: Envelope.SapientSignature = {
+      imageHash,
+      signature,
+    }
+    const signatures: (Envelope.SapientSignature | Envelope.Signature)[] = [sapientSignature]
+
+    // TODO: check whether guard signature is even needed for this wallet based on the topology
+    if (guard) {
+      // Important: guard is not a sapient signer, so we need to unparent the payload
+      const unparentedPayload = {
+        ...payload,
+        parentWallets: undefined,
+      }
+      const digest = Payload.hash(wallet, chainId, unparentedPayload)
+      const typedData = Payload.toTyped(wallet, chainId, unparentedPayload)
+      const serialized = Hex.fromString(TypedData.serialize(typedData))
+
+      try {
+        const guardSignature = await guard.sign(wallet, chainId, digest, serialized)
+        signatures.push({
+          address: guard.address,
+          signature: {
+            type: 'hash',
+            ...guardSignature,
+          },
+        })
+      } catch (error) {
+        // We don't throw here because we may have reached threshold with the sapient signature alone
+        console.error('failed to sign with guard', error)
+      }
+    }
```

Comment:
I don't think this logic belongs in the session manager. Can we have a guard signer in core that implements these functions instead?
When signing the envelope in the dapp-client, the package can sign with the session manager and then if the signature doesn't met the wallet threshold, pass it to the guard to sign.